recipes/tng/all/conanfile.py

```diff
@@ -0,0 +1,88 @@
+from conans import ConanFile, tools, CMake
+import os
+import shutil
+import glob
+
+
+class tngConan(ConanFile):
+    name = "tng"
+    description = "External GROMACS library for loading tng files."
+    license = "BSD"
+    topics = ("conan", "tng", "gromacs")
+    homepage = "https://github.com/gromacs/tng/"
+    url = "https://github.com/conan-io/conan-center-index"
+    exports_sources = "CMakeLists.txt"
+    generators = "cmake"
+    settings = "os", "arch", "compiler", "build_type"
+    options = {
+        "shared": [True, False],
+        "fPIC": [True, False]
+    }
+    default_options = {
+        "shared": False,
+        "fPIC": True
+    }
+
+    requires = (
+        "zlib/1.2.11"
+    )
+
+    _cmake = None
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    @property
+    def _build_subfolder(self):
+        return "build_subfolder"
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def configure(self):
+        del self.settings.compiler.libcxx
+        del self.settings.compiler.cppstd
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        url = self.conan_data["sources"][self.version]["url"]
+        extracted_dir = self.name + "-" + self.version
+        os.rename(extracted_dir, self._source_subfolder)
+
+    def build(self):
+        cmake = self._configure_cmake()
+        cmake.build()
+
+    def _configure_cmake(self):
+        if self._cmake:
+            return self._cmake
+        self._cmake = CMake(self)
+
+        self._cmake.definitions["TNG_BUILD_OWN_ZLIB"] = False
+
+        self._cmake.configure(build_folder=self._build_subfolder)
```

Comment:
Is your `version.h` workaround still needed if you do not use a `build_folder` (and do a in-tree build)?

Alternatively, I would patch the `CMakeLists.txt`.
Looking at a glance, you can modify `CMAKE_CURRENT_BINARY_DIR` [this line](https://github.com/gromacs/tng/blob/14dd68ff46df1ef1cf2d9bcc1f4f5bddb974ce1c/CMakeLists.txt#L63) with `CMAKE_CURRENT_SOURCE_DIR`.

Reply:
I think this problem is [caused by these lines](https://github.com/gromacs/tng/blob/14dd68ff46df1ef1cf2d9bcc1f4f5bddb974ce1c/BuildTNG.cmake#L2-L6) in `BuildTNG.cmake`:

`CMAKE_SOURCE_DIR` and `CMAKE_BUILD_DIR` should be `PROJECT_SOURCE_DIR` and `PROJECT_BINARY_DIR`.

Reply:
I figured it would be an issue with the project's cmake file. Not sure what I should do about it, should we apply a patch? My preference is to not edit the original if I can avoid it.

Reply:
The change would be equivalent to your move of sources.
The advantage of patching is that, when upstream fixes this problem, the patch can simply be dropped without modifying the recipe.

Reply:
I just tested it locally.
This patch avoid any special actions in the `_configure_cmake` function:
```
--- BuildTNG.cmake
+++ BuildTNG.cmake
@@ -1,9 +1,9 @@
 set(TNG_ROOT_SOURCE_DIR ${CMAKE_CURRENT_LIST_DIR})
-file(RELATIVE_PATH TNG_ROOT_BINARY_DIR ${CMAKE_SOURCE_DIR} ${TNG_ROOT_SOURCE_DIR})
+file(RELATIVE_PATH TNG_ROOT_BINARY_DIR ${PROJECT_SOURCE_DIR} ${TNG_ROOT_SOURCE_DIR})
 if ("${TNG_ROOT_BINARY_DIR}" MATCHES "^\.\.")
     set(TNG_ROOT_BINARY_DIR tng)
 endif()
-set(TNG_ROOT_BINARY_DIR ${CMAKE_BINARY_DIR}/${TNG_ROOT_BINARY_DIR})
+set(TNG_ROOT_BINARY_DIR ${PROJECT_BINARY_DIR}/${TNG_ROOT_BINARY_DIR})
 
 set(TNG_MAJOR_VERSION "1")
 set(TNG_MINOR_VERSION "8")
```
I think upstream would be happy to apply this.
Because right now, it could not be included as a subproject.

Reply:
Okay, I added it as a patch.

Reply:
I've opened an issue with the patch at https://github.com/gromacs/tng/issues/1

Reply:
I'm not sure if anyone is even bothering to look at the issues in the original project. I don't think that library has changed in many years

Reply:
It's up to them to accept or not...